### PR TITLE
FEAT Add MoE-to-Dense adapter

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Implements "Pruning and Distilling Mixture-of-Experts into Dense Language Models"

https://huggingface.co/papers/2605.28207

Reference implementation: https://github.com/krafton-ai/moe-to-dense/

## Purpose

Mixture-of-Experts (MoE) models have the advantage that despite having a high parameter count, they are relatively quick at inference because only a sparse subset of "expert" weights (usually, a small subset of weights is chosen on a per-token level). However, they usually still require a lot of memory at inference time, as all experts are held in memory for quick access. The idea of distilling MoE into dense layers is that for specific tasks, we may only need a dense subset of those experts. Think about using a generalist LLM in a very specific domain or a multi-lingual LLM only in one language. In an ideal scenario, after distillation, the model is not only more parameter-efficient, but also quicker at inference.

The paper proposes a kind of self-distillation approach to train a dense FFN to approximate the MoE module. During training, the original model is used as the teacher and the dense path is the student. For inference, the MoE weights are discarded and only the dense FFN is used. The paper tests various different settings and models to find the best working hyper-parameters.

## Description

Technically speaking, this PEFT method does not constitute "parameter-efficient fine-tuning" because at training time, the trained parameters can be quite substantial in size, as we intend to replace the MoE layers with big dense layers (typical they would be something like 25% of total parameters being trained). However, I think this is still a nice fit for PEFT:

- We make the model more parameter-efficient, just not during training but during inference.
- This PEFT method may later be extended to be truly parameter-efficient during training if new ways of efficiently compressing the MoE layers emerge.
- The PEFT ergonomics are quite a nice fit: adding the adapter, handling of multiple adapters, deleting adapters, disabling adapters. Especially the latter comes in quite handy, as we need the teacher during the distillation step, which can be retrieved by disabling the adapter.
- Distilling the MoE weights could benefit from also adapting downstream modules that consume the modified MoE outputs besides other MoE layers. This can be naturally expressed in PEFT via `modules_to_save`, which allows to adapt further layers by fully fine-tuning them (as they are typically small compared to the MoE weights, that is fine).

## Implementation

Although merging is not directly supported, this method offers a conceptually similar method: `peft_model.compress_and_unload`. This works similarly to `peft_model.merge_and_unload` in that it returns a PEFT-free model for inference. However, it throws away the big MoE weights and replaces them with the trained dense FFN. Since this changes the base model architecture, I opted for a different method name here.

Note that for this compressed model to be able to be loaded independent of PEFT, i.e. as a standalone Transformers model, the MoE model must have a dense 'sibling' architecture. This is currently implemented for Qwen3, but other MoE models like GPT-OSS don't have such a sibling. For these models, we create a new model with a single dense expert instead of multiple sparse experts (making a best effort to express this in the model config).

Note that if expressing the dense model in this way fails, there is always the option to load the adapter as PEFT model and call `compress_and_unload` to get the dense variant. However, the corresponding checkpoint won't be loadable by downstream inference engines like llama.cpp or vLLM. If this PEFT method finds adoption, we can work on better supporting more MoE architectures.

(Side note: At one point, I considered adding `num_experts_to_keep_pattern`, analog to `rank_pattern`, which would allow us to have different numbers of experts to keep per layer. However, the resulting unloaded architecture would be impossible to express as a Transformers config, so this option is not included. In an experimental run, it did yield slightly better results with expert numbers chosen heuristically based on calibration data, so this could be a nice feature to add at some point. **Update**: It seems it is possible to express different numbers of experts per layer [via `PretrainedConfig.per_layer_config`](https://huggingface.co/docs/transformers/v5.15.1/en/main_classes/configuration#transformers.PreTrainedConfig.per_layer_config) so this could be a nice future to add. I haven't tested it yet.)

As expert modules are much less standardized than the usual PEFT target modules, some architecture-specific code must exist to accommodate that. This means that if new popular MoE models are added to Transformers, our code might need updating. Special care was taken to provide fallbacks and checks to raise or warn early of expectations are not met. Generally, this PEFT method will most likely not work with models that are not from Transformers.

As for testing, since this model is a bit different from other PEFT methods (no zero init, required calibration step) and strictly requires MoE models as base models, it cannot be included in the normal test matrix. Instead, a separate `tests/test_moe_to_tense.py` was introduced which covers most of what the normal tests would cover.

The implementation was mostly written by Fable with review and heavy editing from my side. Note that the original repo did not implement this in terms of PEFT and is also based on old Transformers v4, so this implementation is not directly comparable.

## Scope

This implementation covers one of the paper's better-performing configuration but not the full breadth of tested options: experts are scored by conditional probability (CP, Section 3.2, Eq. 3), computed purely from router outputs collected during ordinary forward passes. The activation-based scorings (ACP, DO-ACP) that need a separate calibration procedure over expert outputs are not implemented. The top-K experts per layer are then concatenated into a single dense FFN (Section 3.1, Algorithm 1) in the "pure pruning" setting `K = k` that Section 4.2 found best, meaning each kept expert is copied verbatim and no experts are merged into groups (so `num_experts_to_keep` changes the dense FFN's width rather than triggering the paper's merging scheme). The token-dependent routing weights are replaced by the uniform 1/K down-projection scaling of Section 3.4/Appendix H, which makes the conversion exact in the sense of Appendix A.

Refinement uses the paper's winning distillation objective only: forward KL divergence on the output logits (Section 3.5, Eq. 5) with temperature 1 (Appendix J). Reverse KL and the intermediate hidden-state loss from the ablations are not implemented, and the expanded-teacher-routing variant (Eq. 6) is only reachable indirectly by passing precomputed `teacher_logits`.

If this PEFT method finds wider adoption or new results for MoE-to-Dense distillation are published, we can extend the PEFT method to offer more options. The implementation is written in a way that kept future extensions in mind, which means it should (hopefully) be easy to add new options later on.

Also note that some care was taken to allow loading the trained model with MoE layers on meta device. This is so that we can create the compressed model, if needed, without having to actually load the expensive MoE weights. However, PEFT provides no helper function to perform actually this efficient loading.

## Experimental results

To test how this works, a training script is included that is inspired by the MetaMathQA benchmark. I did some smaller tests based on `"ibm-granite/granite-3.1-1b-a400m-instruct"` (see `examples/moe_to_dense/`), as this can be comfortably fitted with 24 GB VRAM.

In these experiments, we can see that the training and test loss decrease over time. Additionally, since the dataset is MetaMathQA/GSM8K, we can also verify the test accuracy. For the base model, we get 28.7% and for the compressed model 10-11% with the given settings. This is of course still a big performance gap, but since the loss is still declining at that point, presumably longer training can recover more of the difference. The paper reports training on the billion parameter range for good results, which I did not test.

To demonstrate the gains from the dense model conversion, here are some numbers with `num_experts_to_keep=4` (original: 8 active experts) after 10,000 training stes:

```
                            MoE (original)   dense (distilled)
parameters (B)                        1.33                0.28
weight memory (GB)                    2.49                0.52
prefill (tokens/s)                45955.71            76902.32
decode bs=1 (ms/step)                10.93               10.57
peak memory bs=1 (GB)                 3.94                1.96
decode bs=16 (ms/step)               29.12               14.46
peak memory bs=16 (GB)                4.50                2.34
```

We can thus expect quite a nice saving in VRAM and also in runtime. Note, however, that the paper itself reports that a dense model of similar size that was fully pre-trained will generally beat the distilled dense model. Still, for cases where such a dense model does not exist, using this PEFT method can be helpful.

### CODE REVIEW

File: src/peft/tuners/moe_to_dense/export.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+        if not verified:
+            # Compare the output of the whole MoE block before and after the replacement (once, as all layers share
+            # the same structure). This catches MoE state that lives outside of the router and the experts.
+            dense_param = layer.moe_to_dense_experts[adapter_name].down_proj
+            x = torch.randn(1, 4, layer.layout.hidden_size, device=dense_param.device, dtype=dense_param.dtype)
+            expected = _try_forward(parent, x)
+
+        dense = _prepare_dense_module(layer, adapter_name, config)
+        setattr(parent, target_name, dense)
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/moe_to_dense/export.py:251]: I don' think we can easily make the whole export atomic. My solution is now to catch errors on the export call in `compress_and_unload` and raise an error with a clear message that the model could be in a broken state.

File: src/peft/tuners/moe_to_dense/layer.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+            return self.base_layer(hidden_states, *args, **kwargs)
+
+        dense = self.moe_to_dense_experts[adapter_name]
+        shape = hidden_states.shape
+        hidden_states = hidden_states.reshape(-1, shape[-1])
+        num_tokens = hidden_states.shape[0]
+        # the single dense "expert" is always selected with routing weight 1
+        top_k_index = torch.zeros((num_tokens, 1), dtype=torch.long, device=hidden_states.device)
+        top_k_weights = torch.ones((num_tokens, 1), dtype=hidden_states.dtype, device=hidden_states.device)
+        return dense(hidden_states, top_k_index, top_k_weights).reshape(shape)
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/moe_to_dense/layer.py:324]: No, I think this is okay as is: Your suggestion would require us to disentangle which `args` and `kwargs` correspond to `top_k_index` and `top_k_weights`, which is fiddly. Instead, we make the assumption here that only these three arguments are needed. I added a check to raise a `RuntimeError` when our assumptions don't match.

File: docs/source/package_reference/moe_to_dense.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+- The adapter, i.e. the dense FFNs, can be saved with `save_pretrained` and loaded with [`PeftModel.from_pretrained`] like any other adapter. A loaded adapter does not need calibration. Be aware that the adapter checkpoint is large relatively large compared to a typical PEFT adapter, as it contains large linear layer weights.
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply docs/source/package_reference/moe_to_dense.md]: Done

File: src/peft/tuners/moe_to_dense/model.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+        for _, layer in layers:
+            layer.remove_router_hook()
+
+        exports = {layer.spec.export for _, layer in layers}
+        if len(exports) != 1:
+            raise ValueError(f"All MoE layers must use the same export strategy, got {sorted(exports)}.")
+
+        export = EXPORT_STRATEGIES[exports.pop()]
+        export(self.model, layers, adapter_name, layers[0][1].spec, progressbar)
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/moe_to_dense/model.py]: Done. `spec` is no longer being passed to the export function, instead the export function uses the `spec` from the passed layers.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the review, I addressed your inline comments.
  
> * `test_experts_implementations_agree` skips `grouped_mm` for GPT-OSS; if transformers later adds support, the skip will silently persist — a comment noting to revisit would help, though the existing skip message is clear enough.
> * Several tests depend on very recent transformers classes (`Gemma4ForCausalLM`, `NemotronHExperts`, `DeepseekV3Config`); worth a sanity check that the minimum transformers pin in `setup.py` covers these, since the repo supports transformers >= ~4.33 per the contribution guide.

I think those concerns are not very pressing, I didn't address them.

**Comment by jhyuckkim** [comment]
Hi @BenjaminBossan! Thank you so much for building this. I went through the full diff carefully, and it generally looks good to me.

I acknowledge that ACP and DO-ACP scoring methods are intentionally left out, which I think is a reasonable scope for now. If you decide to add them later, I would be glad to help. The missing ingredient is a single calibration pass that records the per-expert
mean output norm and the Gram matrix G_ij = E_t[<f_i(t), f_j(t)>], which can reuse the existing batched experts module on a modest token subsample, and the greedy log-det selection itself is cheap and runs on CPU in seconds.

Other than that, a few minor, non-blocking observations:

1. Routing statistics are shared per layer and never reset between adapters. If a second adapter is added after the first one is allocated and forwards are then run with the first adapter active, the re-registered hook collects statistics on hidden states already shaped by the first adapter's dense layers, and those get baked into the second adapter's allocation. Resetting the stats in `update_layer` for a new adapter, or a note in the docs, would cover it.
2. In `export_one_expert`, the pre-replacement reference output for the equivalence check is computed after the config was already patched. The currently supported blocks cache everything at init so nothing breaks today, but computing the reference before `_patch_moe_config` would be safer for architectures that read config values at forward time.

Thanks again for the excellent work, glad to see the method land in PEFT!

**Comment by BenjaminBossan** [comment]
Thanks for checking @jhyuckkim.

> The missing ingredient is a single calibration pass that [...]

Yes, I expect it shouldn't be too difficult to add. The reason why I left it out for now is to keep the already complex code from growing even more complexity. Adding this later should be fine.

> Other than that, a few minor, non-blocking observations:

Thanks for catching these issues, they should now be fixed:

1. Each adapter now gets their own stats (and associated hooks), preventing contamination.
2. Moved the patch of the config to _after_ determining the reference output.

**Comment by BenjaminBossan** [comment]
_not stale_
